rewriter/util/Util.cc

```diff
@@ -323,4 +323,72 @@ ast::ExpressionPtr ASTUtil::thunkBody(core::MutableContext ctx, ast::ExpressionP
     return std::move(block->body);
 }
 
+namespace {
+
+bool validAttrName(std::string_view name) {
+    return !name.empty() && (isalpha(name.front()) || name.front() == '_') &&
+           absl::c_all_of(name, [](char c) { return isalnum(c) || c == '_'; });
+}
+
+} // namespace
+
+pair<core::NameRef, core::LocOffsets> ASTUtil::getAttrName(core::MutableContext ctx, core::NameRef attrFun,
+                                                           const ast::ExpressionPtr &name) {
+    core::LocOffsets loc;
+    core::NameRef res;
+    if (auto *lit = ast::cast_tree<ast::Literal>(name)) {
+        if (lit->isSymbol()) {
+            res = lit->asSymbol();
+            auto shortName = res.shortName(ctx);
+            if (!validAttrName(shortName)) {
+                if (auto e = ctx.beginIndexerError(name.loc(), core::errors::Rewriter::BadAttrArg)) {
+                    if (shortName.empty()) {
+                        e.setHeader("Attribute names must be non-empty");
+                    } else {
+                        e.setHeader("Bad attribute name `{}`", shortName);
+                    }
+                }
+                return make_pair(core::NameRef::noName(), lit->loc);
+            }
+
+            loc = lit->loc;
+            ENFORCE(ctx.locAt(loc).exists());
+            ENFORCE(ctx.locAt(loc).source(ctx).value().size() > 1 && ctx.locAt(loc).source(ctx).value()[0] == ':');
+            loc = core::LocOffsets{loc.beginPos() + 1, loc.endPos()};
+        } else if (lit->isString()) {
+            res = lit->asString();
+            auto shortName = res.shortName(ctx);
+            if (!validAttrName(shortName)) {
+                if (auto e = ctx.beginIndexerError(name.loc(), core::errors::Rewriter::BadAttrArg)) {
+                    if (shortName.empty()) {
+                        e.setHeader("Attribute names must be non-empty");
+                    } else {
+                        e.setHeader("Bad attribute name `{}`", shortName);
+                    }
+                }
+                return make_pair(core::NameRef::noName(), lit->loc);
```

Comment:
out of curiosity, why not pull all the error handling into the new helper function too, and change the method to `validateAttrArg`?

Reply:
That's a good idea. At the time I was imagining that it might be worth having slightly different messages depending on the type, but as they're the same we shouldn't duplicate code.